src/xkbcomp/keymap-dump.c

```diff
@@ -87,16 +87,24 @@ check_write_buf(struct buf *buf, const char *fmt, ...)
     size_t available;
 
     available = buf->alloc - buf->size;
+    /* Try to write to the buffer and get the required min size (without the
+     * final NULL char).
+     * C11 states that if `available` is 0, then the pointer may be null. */
+    assert(buf->buf || available == 0);
     va_start(args, fmt);
-    printed = vsnprintf(buf->buf + buf->size, available, fmt, args);
+    printed = vsnprintf(buf->buf ? buf->buf + buf->size : buf->buf, available,
+                        fmt, args);
     va_end(args);
 
-    if (printed < 0)
+    if (printed < 0) {
         goto err;
-
-    if ((size_t) printed >= available)
-        if (!do_realloc(buf, printed))
+    } else if ((size_t) printed >= available) {
+        if (!do_realloc(buf, printed + 1))
             goto err;
+    } else {
+        buf->size += printed;
```

Comment:
Good catch, we were calling vsnprintf again even if it wasn't needed.

Reply:
It’s your previous comment that trigger my investigation. Love team work!